src/libraries/Microsoft.Extensions.Options/gen/Parser.cs

```diff
@@ -0,0 +1,664 @@
+                            GetMinimalFQNWithoutGenerics(validatorType),
+                            keyword,
+                            parents,
+                            false,
+                            modelsValidatorTypeValidates));
+                    }
+                }
+            }
+
+            results.AddRange(_synthesizedValidators.Values);
+            _synthesizedValidators.Clear();
+
+            return results;
+        }
+
+        private static bool IsAllowedKind(SyntaxKind kind) =>
+            kind == SyntaxKind.ClassDeclaration ||
+            kind == SyntaxKind.StructDeclaration ||
+            kind == SyntaxKind.RecordStructDeclaration ||
+            kind == SyntaxKind.RecordDeclaration;
+
+        private static string GetTypeKeyword(ITypeSymbol type)
+        {
+            if (type.IsReferenceType)
+            {
+                return type.IsRecord ? "record class" : "class";
+            }
+
+            return type.IsRecord ? "record struct" : "struct";
+        }
+
+        private static string GetTypeKeyword(TypeDeclarationSyntax type) =>
+            type.Kind() switch
+            {
+                SyntaxKind.ClassDeclaration => "class",
+                SyntaxKind.RecordDeclaration => "record class",
+                SyntaxKind.RecordStructDeclaration => "record struct",
+                _ => type.Keyword.ValueText,
+            };
+
+        private static string GetFQN(ISymbol type)
+            => type.ToDisplayString(SymbolDisplayFormat.FullyQualifiedFormat.WithMiscellaneousOptions(SymbolDisplayMiscellaneousOptions.IncludeNullableReferenceTypeModifier));
+
+        private static string GetMinimalFQN(ISymbol type)
+            => type.ToDisplayString(SymbolDisplayFormat.MinimallyQualifiedFormat.AddGenericsOptions(SymbolDisplayGenericsOptions.IncludeTypeParameters));
+
+        private static string GetMinimalFQNWithoutGenerics(ISymbol type)
+            => type.ToDisplayString(SymbolDisplayFormat.MinimallyQualifiedFormat.WithGenericsOptions(SymbolDisplayGenericsOptions.None));
+
+        /// <summary>
+        /// Checks whether the given validator already implement the IValidationOptions&gt;T&lt; interface.
+        /// </summary>
+        private static bool AlreadyImplementsValidateMethod(INamespaceOrTypeSymbol validatorType, ISymbol modelType)
+            => validatorType
+                .GetMembers("Validate")
+                .Where(m => m.Kind == SymbolKind.Method)
+                .Select(m => (IMethodSymbol)m)
+                .Any(m => m.Parameters.Length == NumValidationMethodArgs
+                    && m.Parameters[0].Type.SpecialType == SpecialType.System_String
+                    && SymbolEqualityComparer.Default.Equals(m.Parameters[1].Type, modelType));
+
+        /// <summary>
+        /// Checks whether the given type contain any unbound generic type arguments.
+        /// </summary>
+        private static bool HasOpenGenerics(ITypeSymbol type, out string genericType)
```

Comment:
nit consider a simpler implementation if it works for this scenario - https://github.com/dotnet/runtime/blob/4198ed7b499048dce2e6a82c02558d701d8dcca7/src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Parser.cs#L1033.

Reply:
Both implementations are almost identical. The one used here is doing extra things handling the cases for `ITypeParameterSymbol` and `IArrayTypeSymbol`. I'll keep what we have.

Reply:
Ok I'll examine whether my implementation needs to change.